pkg/clusterconditions/cache/cache.go

```diff
@@ -0,0 +1,169 @@
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/cluster-version-operator/pkg/clusterconditions"
+)
+
+// MatchResult represents a single Match invocation.
+type MatchResult struct {
+	// When is the completion time of the wrapped Match call.
+	When time.Time
+
+	// Access is the time of the most recent access.
+	Access time.Time
+
+	// Match is whether the result was a match or not.
+	Match bool
+
+	// Error is the error returned by the Match call.
+	Error error
+}
+
+// Cache wraps a cluster condition with caching and throttling.
+type Cache struct {
+	// Condition is the wrapped cluster condition.
+	Condition clusterconditions.Condition
+
+	// LastMatch is the completion time of the most recent Match
+	// call evaluated by the wrapped cluster condition.
+	LastMatch time.Time
+
+	// MinBetweenMatches is the minimum duration before a new
+	// Match call may be evaluated by the wrapped cluster condition.
+	MinBetweenMatches time.Duration
+
+	// MinForCondition is the minimum duration before a new Match
+	// call for a given condition may be evaluated.
+	MinForCondition time.Duration
+
+	// Expiration is the duration a value will be cached before
+	// being evicted for stale-ness.
+	Expiration time.Duration
+
+	// MatchResults holds results of previous match invocations.
+	MatchResults map[string]*MatchResult
+}
+
+// Valid returns an error if the wrapped cluster condition considers
+// the value invalid.
+func (c *Cache) Valid(ctx context.Context, condition *configv1.ClusterCondition) error {
+	return c.Condition.Valid(ctx, condition)
+}
+
+// Match returns the match value from the wrapped cluster condition,
+// possibly from a fresh evaluation, possibly from a local cache.
+func (c *Cache) Match(ctx context.Context, condition *configv1.ClusterCondition) (bool, error) {
+	if c.MatchResults == nil {
+		c.MatchResults = map[string]*MatchResult{}
+	}
+
+	keyBytes, err := json.Marshal(condition)
+	if err != nil {
+		return false, fmt.Errorf("unable to marshal condition to JSON for use as a cache key: %w", err)
+	}
+	key := string(keyBytes)
+
+	now := time.Now()
+
+	defer func() {
+		if result, ok := c.MatchResults[key]; ok {
+			result.Access = now
+		} else {
+			c.MatchResults[key] = &MatchResult{Access: now}
+		}
+	}()
+
+	for key, value := range c.MatchResults {
+		age := now.Sub(value.When)
+		aspect := "result"
+		if value.When.IsZero() {
+			age = now.Sub(value.Access)
+			aspect = "queued request"
+		}
+		if age > c.Expiration {
+			klog.V(4).Infof("pruning %q from the condition cache, as the %s is %s old", key, aspect, age)
+			delete(c.MatchResults, key)
+		}
+	}
+
+	sinceLastMatch := now.Sub(c.LastMatch)
+	if sinceLastMatch <= c.MinBetweenMatches {
+		if result, ok := c.MatchResults[key]; ok && !result.When.IsZero() {
+			return result.Match, result.Error
+		}
+		return false, fmt.Errorf("client-side throttling: only %s has elapsed since the last match call completed for this cluster condition backend", sinceLastMatch)
```

Comment:
This seems like an info message rather than an error message. IIUC there was an issue. Either MatchResults doesn't contain the key and/or the result returned has a `When` time of 0 appearing as if Match has never been called before.

Reply:
this is:

1. Caller made a `Cache.Match` call.
2. We have to throttle the call, because our previous actual evaluation was within `MinBetweenMatches` of now.
3. There is no evaluated entry in our `MatchResults` we can return instead.

So the caller gets a "sorry, we're refusing to look that up for you because of CVO-side throttling" message, instead of "yup, condition applies" or "no, condition doesn't apply".  We are going to be complaining with `Recommended=Unknown` (or at least this particular risk will be unknown for the conditional update), and that seems like it warrants an `error` to me.

Reply:
Bad phrasing on my part. I agree it's an error but what I didn't state clearly is that the current error message sounds informational rather than conveying "something is fouled up". Perhaps add "...but there is no evaluated entry in our MatchResults to return".

Reply:
I've pushed 5c22e968 -> 1ce08e79, adding `this cached cluster condition request has been queued for later execution` to echo the error message we return that the end, if the actuall `Match` call was stolen by a thief condition, and we didn't have anything cached for the requested condition.